Feature: WEB-203 New konviw step to render slides from Sanofi custom slide macro

krystiankedra opened the pull request:
- Handling new slide which will be defined by Two Macros:
- Slide - definition for every slide
- Slide Cover - definition for full deck
- Handling strategy which step should be used

krystiankedra committed:
feat: WEB-203 Created new step for addNewSlides which will be generated by macro slide and macro slide cover

krystiankedra committed:
feat: WEB-203 Fixed lint errors

krystiankedra committed:
feat: WEB-203 Removed content as optional argument

krystiankedra committed:
feat: WEB-203 Applying dynamic Theme specify in Macro SlideCover

krystiankedra committed:
feat: WEB-203 Fix lint

krystiankedra committed:
feat: WEB-203 Fixed variables convenction

krystiankedra committed:
feat: WEB-203 Added test for add new slides

krystiankedra committed:
feat: WEB-203 Fixed lint issue

krystiankedra committed:
feat: WEB-203 Repair tests

krystiankedra committed:
feat: WEB-203 Fixed displaying content from slideSettings and animated paragraphs by value with scope of slide

krystiankedra committed:
feat: WEB-203 Fixed test fixAddNewSlides

krystiankedra committed:
feat: WEB-203 Updated field

krystiankedra committed:
feat: WEB-203 Created animated paragraphs for OL and LI

PiiXiieeS commented on src/proxy-page/steps/addSlides.ts:
```diff
@@ -29,7 +29,7 @@ export default (): Step => (context: ContextService): void => {
         const className = section('body').children().first().attr('class') ?? '';
         // Based on the 'tag' name of the first element we will design the slide format
-        switch (section('body').first().children().get(0).tagName) {
+        switch (section('body').first().children().get(0)?.tagName) {
```
@krystiankedra 
I have added this check to avoid a bug that happened when several horizontal lines where together without any content in between.

PiiXiieeS commented on src/proxy-page/steps/addNewSlides.ts:
```diff
@@ -33,7 +33,7 @@ export default (config: ConfigService, content: Content): Step => (context: Cont
     if (existRealChildrenWithText) {
       const htmlElemenet = $(element);
-      const finalElement = (htmlElemenet && htmlElemenet['0']) ?? htmlElemenet
+      const finalElement = (htmlElemenet && htmlElemenet['0']) ?? htmlElemenet;
```
@krystiankedra 
I found a couple of missing semicolon in your previous commit

PiiXiieeS commented on src/proxy-page/proxy-page.service.ts:
```diff
@@ -41,6 +40,9 @@ import addSlidesJS from './steps/addSlidesJS';
 import fixTableBackground from './steps/fixTableBackground';
+import addSlideTypeByStrategy from './strategySteps/addSlideTypeByStrategy';
+import addSlideContextByStrategy from './strategySteps/addSlideContextByStrategy';
+import fixAddNewSlides from './steps/fixAddNewSlides';
```
This step should be removed because we should fix the macro behavior to get the same output in Confluence. Alternatively we will capture and delete the body in the macro or make the settings macro not rich-text.

PiiXiieeS commented on src/proxy-page/proxy-page.service.ts:
```diff
@@ -114,6 +116,7 @@ export class ProxyPageService {
     }
     fixSVG(this.config)(this.context);
     fixTableBackground()(this.context);
+    fixAddNewSlides()(this.context);
```
Related to previous comment on no need of this fixAddNewSlides step.

PiiXiieeS commented on tests/unit/steps/fixAddNewSlides.spec.ts:
```diff
@@ -0,0 +1,24 @@
+import fixAddNewSlides from '../../../src/proxy-page/steps/fixAddNewSlides';
```
Related to previous comment on no need of this fixAddNewSlides step.

PiiXiieeS commented on src/proxy-page/proxy-page.service.ts:
```diff
@@ -172,9 +173,10 @@ export class ProxyPageService {
     fixFrameAllowFullscreen()(this.context);
     fixSVG(this.config)(this.context);
     fixTableBackground()(this.context);
+    fixAddNewSlides()(this.context);
```
Related to previous comment on no need of this fixAddNewSlides step.

PiiXiieeS commented on src/proxy-page/steps/addNewSlides.ts:
```diff
@@ -0,0 +1,152 @@
+      if (possibleNested) {
+        return value.data && value.data.trim().length > 0;
+      }
+      return value.children.some((node) => node.data && node.data.trim().length > 0);
```
With testing slide deck, I get error _TypeError: Cannot read properties of undefined (reading 'some')_ in this line when
```
/Users/I0389170/Code/konviw/src/proxy-page/steps/addNewSlides.ts:28
 const existRealChildrenWithText = element.children.some((value: cheerio.Node & { data: string; children: { data: string }[] }) => {
                                                    ^
TypeError: Cannot read properties of undefined (reading 'some')
```

fix it adding, but to be properly checked if this affects the final logic:
`return value.children?.some((node) => node.data && node.data.trim().length > 0);`

PiiXiieeS commented on src/proxy-page/steps/addNewSlides.ts:
```diff
@@ -0,0 +1,152 @@
+    }
+  };
+
+  const searchByTagToAssignFragment = (tag: string, possibleNested: boolean) => {
```
`searchByTagToAssignFragment` or `callbackToAssignFragmentClass` should be **awaited** because all the slides after the first one flag as animated are animated, despite of selecting 'no' for the other slides.

PiiXiieeS commented on src/proxy-page/steps/addNewSlides.ts:
```diff
@@ -0,0 +1,152 @@
+      // Add fragment class for each paragraph to apply fade-in animation
+      if (convertSlideFragmentValueToBoolean(slideParagraphAnimation)) {
+        searchByTagToAssignFragment('p', true);
+        searchByTagToAssignFragment('ol', false);
```
We should add also `<ul>` for decisions, 
`searchByTagToAssignFragment('ul', false);`
but then all decisions are animated at once.

PiiXiieeS commented on src/proxy-page/steps/fixAddNewSlides.ts:
```diff
@@ -0,0 +1,15 @@
+import { Step } from '../proxy-page.step';
+import { ContextService } from '../../context/context.service';
+
+export default (): Step => (context: ContextService): void => {
```
This step is not needed.

PiiXiieeS commented on tests/unit/steps/addNewSlides.spec.ts:
```diff
@@ -0,0 +1,41 @@
+        context = moduleRef.get<ContextService>(ContextService);
+        config = moduleRef.get<ConfigService>(ConfigService);
+    });
+
```
I guess we are missing some testing cases:

- Add slides
- Add slides with vertical slides
- Display slides over legacy slides via page properties
- Add slides with background image
- Add slides with background video
- Add slides with fragments
- Add slides with special styles (cover, default and blank)
- Add slides with different transitions

PiiXiieeS reviewed (changes requested):
A few changes to perform but the overall job done is really amazing!

krystiankedra committed:
feat: WEB-203 Fixed issue related to comments

krystiankedra committed:
feat: WEB-203 Fixed lint

krystiankedra committed:
feat: WEB-203 Sync branch

krystiankedra committed:
feat: WEB-203 Fixed

krystiankedra committed:
feat: WEB-203 fixed problem for animation list item element

krystiankedra committed:
feat: WEB-203 Added tests for add new slides

krystiankedra committed:
feat: WEB-203 fix test

krystiankedra committed:
feat: WEB-203 Fixed tests

PiiXiieeS committed:
feat: adding 'pre' and 'span' as fragment blocks

- so code blocks and colored paragraphs are also fragments

contributes to WEB-203

krystiankedra committed:
feat: WEB-203 Added missing animation for elements

PiiXiieeS commented on src/proxy-page/steps/addNewSlides.ts:
```diff
@@ -0,0 +1,182 @@
+  const searchByTagToAssignFragment = (expression: string, possibleNested: boolean, slideProperties: cheerio.Element, unexpectedExpressionForNestedElements: string[], forceAssign: boolean) => {
+    $(slideProperties).find(expression).each((_: number, element: cheerio.Element) => {
+      if (possibleNested) {
+        const isCorrectConditionalToAssignFragment = unexpectedExpressionForNestedElements.some((expression) => $(element).parents(expression)?.length > 0);
```
This line has a length of 156. Maximum allowed is 140.eslint[max-len](https://eslint.org/docs/rules/max-len)

'expression' is already declared in the upper scope on line 62 column 40.eslint[@typescript-eslint/no-shadow](https://typescript-eslint.io/rules/no-shadow)
(parameter) expression: string

PiiXiieeS commented on src/proxy-page/steps/addNewSlides.ts:
```diff
@@ -0,0 +1,182 @@
+    }
+  };
+
+  const searchByTagToAssignFragment = (expression: string, possibleNested: boolean, slideProperties: cheerio.Element, unexpectedExpressionForNestedElements: string[], forceAssign: boolean) => {
```
This line has a length of 193. Maximum allowed is 140.eslint[max-len](https://eslint.org/docs/rules/max-len)

PiiXiieeS committed:
Merge branch 'main' into feature/web-203

PiiXiieeS reviewed (changes requested):
Take a look to the changes proposed by my linter.

krystiankedra committed:
feat: WEB-203 Fixed lint issues
